### ISSUE TITLE
Namespace Classes

### ISSUE THREAD

**Issue by Zauberfisch** [swipestripe/silverstripe-swipestripe#32]
I noticed many classes in SwipeStripe are very generic.
- Item
- ItemOption
- Modification
- Order
- GridFieldConfig_Basic
- GridFieldDetailForm_HasManyItemRequest
- OptionField
- Price
- PriceField
- AccountPage
- Customer
- NotificationEmail
- Payment_Extension
- Attribute
- Option 
- Variation

Those names are likely to be used in another module as well. Some more than others, but if just one is the same in another module, that makes them instantly incompatible.

It is important that we somehow namespace those items.
One way of doing it would be to rename them to eg `SwipeStripePrice`, but the cleaner way of course would be to use actual namespaces, but unfortunately I don't know what the state on those is. Perhaps @simonwelsh can bring us up to speed on the status of namespace support in SilverStripe (pretty please)?

**Comment by icecaster** [comment]
Hey
Back in the office now, taking part in these discussions now ;)

Agree with you. Frank & I had a discussion about this a while back -
We sort of agreed on using SWS as a prefix for namespacing - however implementing this would be quite a big job, and also require a sophisticated migration script / make upgrades difficult..
So unfortunately it never became a priority for us.
Also, we were, of course, hoping for a cleaner implementation, using php namespaces...

Cheers

**Comment by frankmullenger** [comment]
Yes, totally agree. I think we should use PHP namespaces and last time I looked (if I understood it correctly) you could import namespaces meaning that most of the code wouldn't change. The files would still know a Product as Product rather than having to search and replace with \SwipeStripe\Product everywhere. The backslashes still feel a bit yuck.

**Comment by willmorgan** [comment]
FWIW, I've tried using namespaces with SilverStripe and it's a nightmare. Your problems will start with conflicting classes, and when you try to qualify a namespaced class, it won't get picked up by the `ClassLoader`.

Also, template file naming conventions would need to be updated...

**Comment by simonwelsh** [comment]
I've had no problems at all using namespaces. You just can't use imports for the superclass or interfaces. Apart from that, one namespace per file (using the `namespace blah;` syntax) and it all works.

**Comment by stevie-mayhew** [comment]
So this is legitimate:

``` php
namespace SwipeStripe;

class SwipeStripePage extends \Page 
{
}
```

``` php
namespace SwipeStripe;

class OrderPage extends SwipeStripePage 
{
}
```

and this is not:

``` php
use SwipeStripe\SwipeStripePage;

class OrderPage extends SwipeStripePage 
{
}
```

?

**Comment by simonwelsh** [comment]
Yes.

**Comment by Zauberfisch** [comment]
thanks @simonwelsh!

that sounds like the php part works, but will we not run into any troubles with database tables?
let pick for example class `Item extends DataObject`. we would turn that into `SwipeStripe\Item extends DataObject`, but what happens with the database table? and what happens if there is another module with a dataobject Item?

**Comment by simonwelsh** [comment]
The table is `SwipeStripe\Item`, so there's no collision with other modules.

**Comment by Zauberfisch** [comment]
wonderful.

@stevie-mayhew got another point just a minute ago on IRC:

> I'm also wondering what happens to the DataExtension availability when you namespace

are there any problems to be expected there?

**Comment by simonwelsh** [comment]
Nope. You just need to rename to quote keys/value in the YAML, like https://svn.pocketrent.com/public/oauth/trunk/_config/injector.yml

**Comment by Zauberfisch** [comment]
sounds like the namespace support is a lot better than I expected. 
thanks again @simonwelsh, and I guess further thanks are in order since you are probably responsible for the namespace support.

**Comment by stevie-mayhew** [comment]
Brilliant - thanks for your help @simonwelsh! Is there any more documentation about it or just in your brain?

**Comment by frankmullenger** [comment]
Cheers @simonwelsh! Sounds like namespaces could be a good option.

**Comment by frankmullenger** [comment]
After discussing with Hamish and Damian, perhaps makes sense to look further at namespace support when the core has a namespace solution defined - for migration, table names etc.

**Comment by stevie-mayhew** [comment]
Yeah I agree with that - we don't want to reinvent the wheel over and over.
